# Lab 5 - Neural Networks

We will start thios lab by importing the data and visualizing the data in a summary function and a head of the dataset.

In [104]:
import pandas as pd 

#Get multiple outputs in one single codeblock
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"

df = pd.read_csv("/Users/jaredmcmullen/Desktop/GSB-S545/data/income_evaluation.csv")
df.describe()
df

age        fnlwgt   education-num   capital-gain  \
count  32561.000000  3.256100e+04    32561.000000   32561.000000   
mean      38.581647  1.897784e+05       10.080679    1077.648844   
std       13.640433  1.055500e+05        2.572720    7385.292085   
min       17.000000  1.228500e+04        1.000000       0.000000   
25%       28.000000  1.178270e+05        9.000000       0.000000   
50%       37.000000  1.783560e+05       10.000000       0.000000   
75%       48.000000  2.370510e+05       12.000000       0.000000   
max       90.000000  1.484705e+06       16.000000   99999.000000   

        capital-loss   hours-per-week  
count   32561.000000     32561.000000  
mean       87.303830        40.437456  
std       402.960219        12.347429  
min         0.000000         1.000000  
25%         0.000000        40.000000  
50%         0.000000        40.000000  
75%         0.000000        45.000000  
max      4356.000000        99.000000

age          workclass   fnlwgt    education   education-num  \
0       39          State-gov    77516    Bachelors              13   
1       50   Self-emp-not-inc    83311    Bachelors              13   
2       38            Private   215646      HS-grad               9   
3       53            Private   234721         11th               7   
4       28            Private   338409    Bachelors              13   
...    ...                ...      ...          ...             ...   
32556   27            Private   257302   Assoc-acdm              12   
32557   40            Private   154374      HS-grad               9   
32558   58            Private   151910      HS-grad               9   
32559   22            Private   201490      HS-grad               9   
32560   52       Self-emp-inc   287927      HS-grad               9   

            marital-status          occupation    relationship    race  \
0            Never-married        Adm-clerical   Not-in-family   White   
1       Married-civ-spouse     Exec-managerial         Husband   White   
2                 Divorced   Handlers-cleaners   Not-in-family   White   
3       Married-civ-spouse   Handlers-cleaners         Husband   Black   
4       Married-civ-spouse      Prof-specialty            Wife   Black   
...                    ...                 ...             ...     ...   
32556   Married-civ-spouse        Tech-support            Wife   White   
32557   Married-civ-spouse   Machine-op-inspct         Husband   White   
32558              Widowed        Adm-clerical       Unmarried   White   
32559        Never-married        Adm-clerical       Own-child   White   
32560   Married-civ-spouse     Exec-managerial            Wife   White   

           sex   capital-gain   capital-loss   hours-per-week  native-country  \
0         Male           2174              0               40   United-States   
1         Male              0              0               13   United-States   
2         Male              0              0               40   United-States   
3         Male              0              0               40   United-States   
4       Female              0              0               40            Cuba   
...        ...            ...            ...              ...             ...   
32556   Female              0              0               38   United-States   
32557     Male              0              0               40   United-States   
32558   Female              0              0               40   United-States   
32559     Male              0              0               20   United-States   
32560   Female          15024              0               40   United-States   

       income  
0       <=50K  
1       <=50K  
2       <=50K  
3       <=50K  
4       <=50K  
...       ...  
32556   <=50K  
32557    >50K  
32558   <=50K  
32559   <=50K  
32560    >50K  

[32561 rows x 15 columns]

Next, we will further dig into the data and determine if there are any duplicates, NAs, and determine if the dataset is balanced or unbalanced.

## Preprocessing 

### Preprocessing - Exploration

In [105]:
#View the number of duplicates in columns
duplicates = df[df.duplicated()]
dup_shape = duplicates.shape

#len of income values in dataset
lessthan50k = len(df[df[' income'] == ' <=50K'])
greaterthan50k = len(df[df[' income'] == ' >50K'])
lengthdf = len(df)

print("---------------------------------------")
print("Amount of NA values in the dataset:\n", df.isna().sum())
print("---------------------------------------")
print("Amount of duplicate rows: ", dup_shape[0]  )
print("---------------------------------------")
print("Length of Dataset ", lengthdf)
print("---------------------------------------")
print("<=50k Customer Count: ", lessthan50k)
print("---------------------------------------")
print(">50k Customer Count: ", greaterthan50k)
print("---------------------------------------")
print("Ratio of dataset", lessthan50k/lengthdf, greaterthan50k/lengthdf )

---------------------------------------
Amount of NA values in the dataset:
 age                0
 workclass         0
 fnlwgt            0
 education         0
 education-num     0
 marital-status    0
 occupation        0
 relationship      0
 race              0
 sex               0
 capital-gain      0
 capital-loss      0
 hours-per-week    0
 native-country    0
 income            0
dtype: int64
---------------------------------------
Amount of duplicate rows:  24
---------------------------------------
Length of Dataset  32561
---------------------------------------
<=50k Customer Count:  24720
---------------------------------------
>50k Customer Count:  7841
---------------------------------------
Ratio of dataset 0.7591904425539756 0.2408095574460244


In [107]:
for col in df:
    print(df[col])
    print(df[col].unique())


0        39
1        50
2        38
3        53
4        28
         ..
32556    27
32557    40
32558    58
32559    22
32560    52
Name: age, Length: 32561, dtype: int64
[39 50 38 53 28 37 49 52 31 42 30 23 32 40 34 25 43 54 35 59 56 19 20 45
 22 48 21 24 57 44 41 29 18 47 46 36 79 27 67 33 76 17 55 61 70 64 71 68
 66 51 58 26 60 90 75 65 77 62 63 80 72 74 69 73 81 78 88 82 83 84 85 86
 87]
0                State-gov
1         Self-emp-not-inc
2                  Private
3                  Private
4                  Private
               ...        
32556              Private
32557              Private
32558              Private
32559              Private
32560         Self-emp-inc
Name:  workclass, Length: 32561, dtype: object
[' State-gov' ' Self-emp-not-inc' ' Private' ' Federal-gov' ' Local-gov'
 ' ?' ' Self-emp-inc' ' Without-pay' ' Never-worked']
0         77516
1         83311
2        215646
3        234721
4        338409
          ...  
32556    257302
32557    154374
32558 

### Preprocessing - Changing the data

From the preprocessing above, we can see that the dataset is unbalanced, skewed towards individuals with less than 50k in salary. So in order to yield the best results, we will balance the dataset.

In [112]:
# Create function to balance the dataset

def downsample(df:pd.DataFrame, label_col_name:str) -> pd.DataFrame:
    # find the number of observations in the smallest group
    nmin = df[label_col_name].value_counts().min()
    return (df
            # split the dataframe per group
            .groupby(label_col_name)
            # sample nmin observations from each group
            .apply(lambda x: x.sample(nmin))
            # recombine the dataframes
            .reset_index(drop=True)
            )
df_balanced = downsample(df, " income")


# Rerun test to see balanced dataset
lessthan50k = len(df_balanced[df_balanced[' income'] == ' <=50K'])
greaterthan50k = len(df_balanced[df_balanced[' income'] == ' >50K'])
lengthdf = len(df_balanced)

print("---------------------------------------")
print("<=50k Customer Count: ", lessthan50k)
print("---------------------------------------")
print(">50k Customer Count: ", greaterthan50k)
print("---------------------------------------")
print("Ratio of dataset", lessthan50k/lengthdf, greaterthan50k/lengthdf )

---------------------------------------
<=50k Customer Count:  7508
---------------------------------------
>50k Customer Count:  7508
---------------------------------------
Ratio of dataset 0.5 0.5


In [113]:
#Get rid of all of the ? values in the dataset
df.drop(df.index[df[' workclass'] == ' ?'], inplace = True)
df.drop(df.index[df[' occupation'] == ' ?'], inplace = True)
df.drop(df.index[df[' native-country'] == ' ?'], inplace = True)

### Preprocessing - Label Encoder

In [114]:
from sklearn.preprocessing import LabelEncoder

#Use label encoder to get in the model ingesting format
for label in df_balanced.columns:
    df_balanced[label] = LabelEncoder().fit(df_balanced[label]).transform(df_balanced[label])

#Set X and Y values for the models
Y = df_balanced[' income']
X = df_balanced.drop([' income'],axis=1)
print("X", "----------------------")
X
print("Y", "----------------------")
Y

X ----------------------


age   workclass   fnlwgt   education   education-num   marital-status  \
0        0           2     6734           1               6                4   
1       46           2     9144          11               8                2   
2        7           2     7170           7              11                4   
3       19           2      681          15               9                6   
4        1           2     9361          15               9                4   
...    ...         ...      ...         ...             ...              ...   
15011   48           2     6524           9              12                2   
15012   29           2     4745          15               9                2   
15013   34           1     5436           9              12                2   
15014   29           3     9230           9              12                2   
15015   35           2    10867           8              10                2   

        occupation   relationship   race   sex   capital-gain   capital-loss  \
0               11              3      4     0              0              0   
1                0              5      4     0             46              0   
2                9              3      2     1              0              0   
3               11              1      4     0              0              0   
4                0              1      4     1              0              0   
...            ...            ...    ...   ...            ...            ...   
15011            2              0      4     1              0              0   
15012           13              0      4     1              0              0   
15013            9              0      4     1              0             40   
15014           11              0      4     1              0              0   
15015            3              0      4     1              0              0   

        hours-per-week   native-country  
0                   14               38  
1                   39               38  
2                   19               38  
3                   19               38  
4                   39               38  
...                ...              ...  
15011               39                8  
15012               44               38  
15013               39               38  
15014               68               38  
15015               39               38  

[15016 rows x 14 columns]

Y ----------------------


0        0
1        0
2        0
3        0
4        0
        ..
15011    1
15012    1
15013    1
15014    1
15015    1
Name:  income, Length: 15016, dtype: int64

## Neural Network Modeling

Now that our data is prepped, we can begin the modeling process. We will start this process by determining the optimal neural network followed by comparing these accuracy scores 

In [115]:
#Create a function to get an accuracy score 
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.ensemble import BaggingRegressor
from sklearn.model_selection import RepeatedStratifiedKFold
from sklearn.model_selection import cross_val_score
from numpy import mean
from numpy import std

#Scale the data
from sklearn.preprocessing import StandardScaler 
scaler = StandardScaler() 

#Set up new function to scale the x data
def evaluate_model(model, X, Y):
    scaler.fit(X)
    X_scaled = scaler.transform(X)  
    cv = RepeatedStratifiedKFold(n_splits=10, n_repeats=3, random_state=1)
    scores = cross_val_score(model, X_scaled, Y, scoring='accuracy', cv=cv, n_jobs=-1)
    return scores

### Basic NN

In [116]:
from sklearn.neural_network import MLPClassifier

def get_models():
    models = dict()
    models['nn_basic'] = MLPClassifier()
    return models

models = get_models()

#Iterate through the models and append the scores
from matplotlib import pyplot
results, names = list(), list()
for name, model in models.items():
    scores = evaluate_model(model, X, Y)
    results.append(scores)
    names.append(name)
    print('>%s %.3f (%.3f)' % (name, mean(scores), std(scores)))

>nn_basic 0.822 (0.009)


### Grid Search

In [117]:
#GridSearch
from sklearn.model_selection import GridSearchCV
model = MLPClassifier()
parameter_space = {
    'hidden_layer_sizes': [(10,30,10),(20,), (100,)],
    'activation': ['tanh', 'relu', 'identity', 'logistic'],
    'solver': ['sgd', 'adam', 'lbfgs'],
    'alpha': [0.0001, 0.05],
    'learning_rate': ['constant','adaptive', 'invscaling'],
}


clf = GridSearchCV(model, parameter_space, n_jobs=-1, cv=5)
clf.fit(X, Y)

GridSearchCV(cv=5, estimator=MLPClassifier(), n_jobs=-1,
             param_grid={'activation': ['tanh', 'relu', 'identity', 'logistic'],
                         'alpha': [0.0001, 0.05],
                         'hidden_layer_sizes': [(10, 30, 10), (20,), (100,)],
                         'learning_rate': ['constant', 'adaptive',
                                           'invscaling'],
                         'solver': ['sgd', 'adam', 'lbfgs']})

### Final Model

In [118]:
final_model = clf.best_params_
print('Best parameters found:\n', clf.best_params_)

Best parameters found:
 {'activation': 'logistic', 'alpha': 0.0001, 'hidden_layer_sizes': (100,), 'learning_rate': 'constant', 'solver': 'adam'}


### Final Model  - Results

In [119]:
from sklearn.neural_network import MLPClassifier
from sklearn.ensemble import BaggingClassifier

final_model = MLPClassifier(activation = 'logistic', alpha = 0.0001, hidden_layer_sizes = (20,), learning_rate = 'invscaling', solver = 'adam')

def get_models():
    models = dict()
    models['nn_basic'] = final_model
    return models

models = get_models()

#Iterate through the models and append the scores
from matplotlib import pyplot
results, names = list(), list()
for name, model in models.items():
    scores = evaluate_model(model, X, Y)
    results.append(scores)
    names.append(name)
    print('>%s %.3f (%.3f)' % (name, mean(scores), std(scores)))

>nn_basic 0.823 (0.010)


The final tuned model accuracy that we could achieve is 82.1 percent accurate. Up from 82 percent in our basic neural network. The data that we are working with has been label encoded and scaled, which yields much higher accuracy scores. From here, I will explore additional models outside of the neural network makeup. 

## Linear Classifier Models

In [51]:
from sklearn.naive_bayes import GaussianNB
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import BaggingClassifier
from sklearn.ensemble import AdaBoostClassifier

#Supress warnings and set the seed
import warnings
warnings.filterwarnings("ignore")

#Define the models that are going to be iterated through later on 
def get_models():
    models = dict()
    models['lr'] = LogisticRegression()
    models['lr_boost'] = AdaBoostClassifier(base_estimator = LogisticRegression())
    models['lr_bag'] = BaggingClassifier(base_estimator = LogisticRegression())
    models['bayes'] = GaussianNB()
    return models

models = get_models()
#Iterate through the models and append the scores
from matplotlib import pyplot
results, names = list(), list()
for name, model in models.items():
    scores = evaluate_model(model, X, Y)
    results.append(scores)
    names.append(name)
    print('>%s %.3f (%.3f)' % (name, mean(scores), std(scores)))

>lr 0.770 (0.011)
>lr_boost 0.767 (0.010)
>lr_bag 0.770 (0.011)
>bayes 0.768 (0.011)


Our linear classifier models tend to be all around the same range of .770 regardless of the method used. We can rule linear classifier models out of our modeling efforts and focus on other strategies. 

## Test Basic Random Forest & Decision Tree Models

In [54]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
#Define the models that are going to be iterated through later on 
def get_models():
    models = dict()
    models['rf'] = RandomForestClassifier()
    models['dec_tree'] = DecisionTreeClassifier()
    return models

models = get_models()
#Iterate through the models and append the scores
from matplotlib import pyplot
results, names = list(), list()
for name, model in models.items():
    scores = evaluate_model(model, X, Y)
    results.append(scores)
    names.append(name)
    print('>%s %.3f (%.3f)' % (name, mean(scores), std(scores)))

>rf 0.820 (0.010)
>dec_tree 0.768 (0.011)


The base random forest model appears to be performing strongly with the dataset, so we will continue to explore random forests related to some other methods. 

## Ensemble Models - Bagging

In [56]:
#Define the models that are going to be iterated through later on 
def get_models():
    models = dict()
    models['Dectree_bag'] = BaggingClassifier(base_estimator = DecisionTreeClassifier())
    models['rf_bag'] = BaggingClassifier(base_estimator = RandomForestClassifier())
    return models

models = get_models()
#Iterate through the models and append the scores
from matplotlib import pyplot
results, names = list(), list()
for name, model in models.items():
    scores = evaluate_model(model, X, Y)
    results.append(scores)
    names.append(name)
    print('>%s %.3f (%.3f)' % (name, mean(scores), std(scores)))

>Dectree_bag 0.804 (0.008)
>rf_bag 0.825 (0.009)


Our Random Forest that utilizes bagging yields the best results on the data. We will continue to use random forests on the dataset and try our boosting and stacking. 

## Ensemble Models - Boosting

In [59]:
from xgboost import XGBClassifier
#Define the models that are going to be iterated through later on 
def get_models():
    models = dict()
    models['ada_randomforest'] = AdaBoostClassifier(base_estimator = RandomForestClassifier())
    models['xgb_dectree'] = XGBClassifier()
    return models

models = get_models()
#Iterate through the models and append the scores
from matplotlib import pyplot
results, names = list(), list()
for name, model in models.items():
    scores = evaluate_model(model, X, Y)
    results.append(scores)
    names.append(name)
    print('>%s %.3f (%.3f)' % (name, mean(scores), std(scores)))

>ada_randomforest 0.820 (0.009)
>xgb_dectree 0.836 (0.008)


In [65]:
from xgboost import XGBClassifier
#Define the models that are going to be iterated through later on 
def get_models():
    models = dict()
    models['ada_randomforest'] = AdaBoostClassifier(base_estimator = RandomForestClassifier())
    models['xgb_dectree'] = XGBClassifier()
    return models

models = get_models()
models

{'ada_randomforest': AdaBoostClassifier(base_estimator=RandomForestClassifier()),
 'xgb_dectree': XGBClassifier(base_score=None, booster=None, callbacks=None,
               colsample_bylevel=None, colsample_bynode=None,
               colsample_bytree=None, early_stopping_rounds=None,
               enable_categorical=False, eval_metric=None, gamma=None,
               gpu_id=None, grow_policy=None, importance_type=None,
               interaction_constraints=None, learning_rate=None, max_bin=None,
               max_cat_to_onehot=None, max_delta_step=None, max_depth=None,
               max_leaves=None, min_child_weight=None, missing=nan,
               monotone_constraints=None, n_estimators=100, n_jobs=None,
               num_parallel_tree=None, predictor=None, random_state=None,
               reg_alpha=None, reg_lambda=None, ...)}

XGBoosing has yielded the highest accuracy and slow standard deviation from that accuracy. We will continue to explore these models as it relates to stacking.

## Ensemble Models - Stacking

In [96]:
from sklearn.ensemble import StackingClassifier

#Define the stacking model used later on
def get_stacking():
    # define the base models
    level0 = list()
    level0.append(('ada_dec', AdaBoostClassifier(base_estimator = RandomForestClassifier())))
    level0.append(('xgb', XGBClassifier()))
    level0.append(('rf_bag', BaggingClassifier(base_estimator = RandomForestClassifier())))
    level0.append(('rf', RandomForestClassifier()))
    #meta learner model
    level1 = LogisticRegression()
    #stacking ensemble
    model = StackingClassifier(estimators = level0, final_estimator = level1, cv = 5)
    return model

#Define the models that are going to be iterated through later on 
def get_models():
    models = dict()
    models['stacking'] = get_stacking()
    models['xgb_dectree'] = XGBClassifier()
    return models

models = get_models()

#Iterate through the models and append the scores
from matplotlib import pyplot
results, names = list(), list()
for name, model in models.items():
    scores = evaluate_model(model, X, Y)
    results.append(scores)
    names.append(name)
    print('>%s %.3f (%.3f)' % (name, mean(scores), std(scores)))

>stacking 0.837 (0.010)
>xgb_dectree 0.836 (0.008)


As we can see from all of the model outputs for accuracy, stacking yields the highest with a 83.7 percent accuracy. This was done by combining all the best performing models. 

## Summary

To conclude this lab, we explored neural networks, linear classification models, and ensemble methods to try and yield the highest accuracy prediction model. The highest neural network accuracy that was achieved was 82.1 percent accurate on a balanced dataset. The highest other model was stacking with 83.7 percent accuraxy utilizing boosting, bagging and a random forest model. Also, the second highest was XGboosting with 83.6 percent accuracy. This lab shows us that while neural networks are very advanced and useful models, maybe they don't work that well on some forms of datasets. This is especially apparent if the dataset has not been scaled (accuracy was around 60 percent without scaling). It is important to know that there is no free lunch, and trying multiple types of models against certainn types of data is extremely important!